### ISSUE TITLE
Using Pytorch Metric Learning for dual encoder structure for question answering (Q&A) 

### ISSUE THREAD

**Issue by puzzlecollector** [KevinMusgrave/pytorch-metric-learning#629]
Hello @KevinMusgrave 
I was thinking about using the pytorch metric learning library for dual encoder based Q&A model. Initially I was using DPR (dense passage retrieval for open domain question answering: https://arxiv.org/abs/2004.04906) but I wanted to try out using pytorch metric learning so I changed the code as follows: 

```
epochs = 30
params = list(q_encoder.parameters()) + list(p_encoder.parameters()) 
optimizer = AdamW(params, lr=2e-5, eps=1e-8) 
t_total = len(train_dataloader) * epochs 
scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps=int(0.05*t_total), num_training_steps=t_total) 

miner = miners.MultiSimilarityMiner() 
loss_func = losses.MultiSimilarityLoss()

q_encoder.zero_grad() 
p_encoder.zero_grad() 
torch.cuda.empty_cache() 

for epoch_i in tqdm(range(0, epochs), desc="Epochs", position=0, leave=True, total=epochs):
    train_loss = 0 
    q_encoder.train() 
    p_encoder.train() 
    with tqdm(train_dataloader, unit="batch") as tepoch: 
        for step, batch in enumerate(tepoch): 
            batch = tuple(t.to(device) for t in batch) 
            q_input_ids, q_attn_masks, q_labels, p_input_ids, p_attn_masks, p_labels = batch 
            q_embeddings = q_encoder(q_input_ids, q_attn_masks).pooler_output 
            p_embeddings = p_encoder(p_input_ids, p_attn_masks).pooler_output 
            full_embeddings = torch.cat((q_embeddings, p_embeddings), dim=0) 
            full_labels = torch.cat((q_labels, p_labels))  
            shuffled_idx = torch.randperm(full_embeddings.shape[0]) 
            full_embeddings, full_labels = full_embeddings[shuffled_idx], full_labels[shuffled_idx] 
            
            hard_pairs = miner(full_embeddings, full_labels) 
            loss = loss_func(full_embeddings, full_labels, hard_pairs) 
            train_loss += loss.item() 
            loss.backward() 
            torch.nn.utils.clip_grad_norm_(params, 1.0)
            optimizer.step() 
            scheduler.step() 
            q_encoder.zero_grad() 
            p_encoder.zero_grad() 
            tepoch.set_postfix(loss=train_loss / (step+1)) 
            time.sleep(0.1) 
```

Is this the correct way of applying pytorch metric learning? so I have two encoders query encoder `q_encoder` and passage encoder `p_encoder` and because the pytorch metric learning library requires embeddings and labels, I calculated the embeddings separately using each encoder and merged them. The way I gave labels is, because I have paired question-answer dataset, I gave these pairs the same labels. More specifically, this is my dataloader code 

```
class PairData(Dataset):
    def __init__(self, df: pd.DataFrame):
        super(PairData, self).__init__()
        self.data = df
    def __getitem__(self, index):
        return self.data.iloc[index]
    def __len__(self):
        return self.data.shape[0]

class custom_collate(object):
    def __init__(self):
        self.tokenizer = AutoTokenizer.from_pretrained("monologg/kobigbird-bert-base")
        self.chunk_size = 512
    def __call__(self, batch):
        q_input_ids, q_attn_masks, q_labels = [], [], []
        p_input_ids, p_attn_masks, p_labels = [], [], [] 
        ids = 0
        all_queries = [] 
        for idx, row in enumerate(batch):
            query, passage = row[0], row[1] 
            encoded_q = self.tokenizer(query, max_length=self.chunk_size, padding="max_length", truncation=True, return_tensors="pt") 
            encoded_p = self.tokenizer(passage, max_length=self.chunk_size, padding="max_length", truncation=True, return_tensors="pt") 
            q_input_ids.append(encoded_q["input_ids"]) 
            q_attn_masks.append(encoded_q["attention_mask"]) 
            q_labels.append(ids) 
            
            p_input_ids.append(encoded_p["input_ids"]) 
            p_attn_masks.append(encoded_p["attention_mask"]) 
            p_labels.append(ids) 
            ids += 1 
        q_input_ids = torch.stack(q_input_ids, dim=0).squeeze(dim=1)  
        q_attn_masks = torch.stack(q_attn_masks, dim=0).squeeze(dim=1) 
        q_labels = torch.tensor(q_labels, dtype=int)  
        
        p_input_ids = torch.stack(p_input_ids, dim=0).squeeze(dim=1) 
        p_attn_masks = torch.stack(p_attn_masks, dim=0).squeeze(dim=1) 
        p_labels = torch.tensor(p_labels, dtype=int) 
        
        return q_input_ids, q_attn_masks, q_labels, p_input_ids, p_attn_masks, p_labels  

# define dataloader 
train_set = PairData(train_df) 
valid_set = PairData(valid_df) 
collate = custom_collate() 

train_dataloader = DataLoader(train_set, batch_size=16, num_workers=4, collate_fn=collate, shuffle=True) 
valid_dataloader = DataLoader(valid_set, batch_size=16, num_workers=4, collate_fn=collate, shuffle=False) 
```

I would appreciate any advice. Thank you :)

**Comment by puzzlecollector** [comment]
I see. Thanks :)
